sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/AsyncResources.scala

```diff
@@ -0,0 +1,442 @@
+      new SimpleAsyncResult(resultData, metrics)
+    } else {
+      new AsyncResultDecayRelease(resultData, metrics,
+        releaseHook = Option(() => this.callDecayReleaseCallback()))
+    }
+  }
+
+  def call(): AsyncResult[T] = {
+    val startTime = System.nanoTime()
+    val resultData = try {
+      beforeExecuteHook.foreach { hook => hook() }
+      callImpl()
+    } finally {
+      afterExecuteHook.foreach { hook => hook() }
+    }
+    metricsBuilder.setExecutionTimeMs(System.nanoTime() - startTime)
+    buildResult(resultData, metricsBuilder.build())
+  }
+
+  protected var beforeExecuteHook: Option[() => Unit] = None
+  protected var afterExecuteHook: Option[() => Unit] = None
+
+  // Set hooks to be executed right before the task execution.
+  def setBeforeHook(hook: () => Unit): Unit = beforeExecuteHook = Some(hook)
+
+  // Set hooks to be executed right after the task execution.
+  def setAfterHook(hook: () => Unit): Unit = afterExecuteHook = Some(hook)
+
+  /**
+   * This method is called when the required resource has been just acquired from pool.
+   * It can be overridden by subclasses to perform actions right after the acquisition.
+   */
+  def onAcquire(): Unit = {
+  }
+
+  /**
+   * This method is called when the acquired resource has been just released back into pool.
+   * It can be overridden by subclasses to perform actions right after the release.
+   */
+  def onRelease(): Unit = {
+  }
+
+  /**
+   * Returns true if the task should release its resources after completion.
+   * This is useful for tasks that are not long-lived and can release resources
+   * immediately after they finish.
+   */
+  def holdResourceAfterCompletion: Boolean = false
+
+  /**
+   * Guarantees the release callback will NOT be called unless the task holds the resource.
+   * This method is protected not private, so it can be called by subclasses to build own result.
+   */
+  private def callDecayReleaseCallback(): Unit = {
+    require(holdResourceAfterCompletion,
+      "This AsyncRunner does not hold resource after completion")
+    require(execException.isEmpty, "The resource was somehow released forcefully, " +
+        s"caught runtime exception: ${execException.get}")
+    require(releaseResourceCallback != null, "releaseResourceCallback is not registered")
+    require(!hasDecayReleased, "callDecayReleaseCallback has been already called for once")
+    releaseResourceCallback()
+    hasDecayReleased = true
+  }
+
+  private[async] var releaseResourceCallback: () => Unit = _
+
+  private[async] lazy val metricsBuilder = new AsyncMetricsBuilder
+
+  private[async] var execException: Option[Throwable] = None
+
+  private var hasDecayReleased = false
+}
+
+/**
+ * An AsyncRunner that requires no resource limits and has the highest scheduling priority.
+ * Useful for lightweight tasks that should execute immediately without resource constraints.
+ */
+abstract class UnboundedAsyncRunner[T] extends AsyncRunner[T] {
```

Comment:
When the pool resource is exhausted, tasks in the thread pool are all sleeping waiting for resources to free up. Each thread poped an item from the work queue and they are waiting up until the time given as a timeout (or whether the resource shows up before).

If we happen to have a plan where unbounded and bounded runners are used at the same time.. say we are scanning orc and parquet? I am not 100% sure what cases cause this mix, but the comment amount won't apply in those cases => unbounded tasks will be stuck because there are no threads to execute them if the bounded tasks are waiting on resources.